pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -298,63 +350,75 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
+		if strings.HasPrefix(s, serviceAccount.Namespace+"/"+osautil.GetTokenSecretNamePrefix(serviceAccount)) {
+			uncastSecret, _, _ := e.secretsCache.GetByKey(s)
+			if secret := uncastSecret.(*api.Secret); len(secret.Annotations) > 0 {
+				if createdBy, exists := secret.Annotations[api.CreatedByAnnotation]; exists && createdBy == "create_dockercfg_secrets" {
+					tokenSecret = secret
+					break
+				}
+			}
+		}
+	}
+	return tokenSecret
+}
 
 // createTokenSecret creates a token secret for a given service account.  Returns the name of the token
-func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
-	tokenSecret := &api.Secret{
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, bool, error) {
+	tokenSecret := e.getTokenSecret(serviceAccount)
+	if tokenSecret != nil {
+		return tokenSecret, len(tokenSecret.Data[api.ServiceAccountTokenKey]) > 0, nil
+	}
+	tokenSecret = &api.Secret{
 		ObjectMeta: api.ObjectMeta{
 			Name:      secret.Strategy.GenerateName(osautil.GetTokenSecretNamePrefix(serviceAccount)),
 			Namespace: serviceAccount.Namespace,
 			Annotations: map[string]string{
 				api.ServiceAccountNameKey: serviceAccount.Name,
 				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+				api.CreatedByAnnotation:   "create_dockercfg_secrets",
 			},
 		},
 		Type: api.SecretTypeServiceAccountToken,
 		Data: map[string][]byte{},
 	}
 
-	_, err := e.client.Secrets(tokenSecret.Namespace).Create(tokenSecret)
+	glog.V(2).Infof("Creating token secret %q for service account %s/%s", tokenSecret.Name, serviceAccount.Namespace, serviceAccount.Name)
+	token, err := e.client.Secrets(tokenSecret.Namespace).Create(tokenSecret)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
-
-	// now we have to wait for the service account token controller to make this valid
-	// TODO remove this once we have a create-token endpoint
-	for i := 0; i <= tokenSecretWaitTimes; i++ {
-		liveTokenSecret, err2 := e.client.Secrets(tokenSecret.Namespace).Get(tokenSecret.Name)
-		if err2 != nil {
-			return nil, err2
-		}
-
-		if len(liveTokenSecret.Data[api.ServiceAccountTokenKey]) > 0 {
-			return liveTokenSecret, nil
-		}
-
-		time.Sleep(wait.Jitter(tokenSecretWaitInterval, 0.0))
-
-	}
-
-	// the token wasn't ever created, attempt deletion
-	glog.Warningf("Deleting unfilled token secret %s/%s", tokenSecret.Namespace, tokenSecret.Name)
-	if deleteErr := e.client.Secrets(tokenSecret.Namespace).Delete(tokenSecret.Name); (deleteErr != nil) && !kapierrors.IsNotFound(deleteErr) {
-		utilruntime.HandleError(deleteErr)
+	// TODO: (@deads) this will/might bork the cache, but it also prevents the race
+	// condition when we create the secret and trigger the 'Add' watch which will not see
+	// this secret in
+	// e.secretsCache.Add(token)
+	//
+	// So instead do polling
+	if err := wait.PollImmediate(10*time.Millisecond, 2*time.Second, func() (bool, error) {
```

Comment:
I'd rather see the controller remember the service accounts it created tokens for. As soon as the token add event is received in the watch, the service account name can be removed from that list. The first part of this function could change to:

```
switch tokenSecret := e.getTokenSecret(serviceAccount) {
case nil:
  if hasPendingToken(serviceAccount) {
    return nil, false, nil
  }
default:
  return tokenSecret, len(tokenSecret.Data[api.ServiceAccountTokenKey]) > 0, nil
}
```